File: 1. Spark Familiarization/submission/task2.py
import sys
import json
import re
from pyspark import SparkContext
sc = SparkContext.getOrCreate()

if __name__ == '__main__':
    spark = True
    if sys.argv[4] == "spark":   
        review = sc.textFile(sys.argv[1])
        business = sc.textFile(sys.argv[2])
    else:
        spark = False
        with open(sys.argv[1],"r") as file:
            review = list(map(lambda x: json.loads(x),file.readlines()))
        with open(sys.argv[2],"r") as file:
            business= list(map(lambda x: json.loads(x),file.readlines()))
 

    output = sys.argv[3]
    stopwords = []
    n = int(sys.argv[5])
    ans = dict()
    
    if spark:
        b = business.map(lambda x:json.loads(x)).filter(lambda x:x['categories'] != None).map(lambda x:(x['business_id'],[i.strip() for i in x['categories'].split(',')]))
        r = review.map(lambda x:json.loads(x)).map(lambda x:(x["business_id"],x["stars"]))
        ans['result'] = b.join(r).flatMap(lambda x:[(i,x[1][1]) for i in x[1][0]]).aggregateByKey((0,0),lambda x,y:(x[0]+y,x[1]+1),lambda x,y:(x[0]+y[0],x[1]+y[1])).mapValues(lambda x:float(x[0]/x[1])).sortBy(lambda x:(-x[1],x[0])).take(n)
    else:

        review_map = {}

        for i in review:
            if i["business_id"] in review_map:
                review_map[i["business_id"]][0]+= i["stars"]
                review_map[i["business_id"]][1]+= 1
            else:
                review_map[i["business_id"]]=[i["stars"],1]
        
        category_map = {}
        for i in business:
            if i["business_id"] not in review_map or i['categories']==None:
                #print(i["business_id"]," not in review")
                continue
            for j in i['categories'].split(','):
                j = j.strip()
                if j not in category_map:
                    category_map[j] = [0,0]
                category_map[j][0] += review_map[i['business_id']][0]
                category_map[j][1] += review_map[i['business_id']][1]
        #print(category_map)
        for i in category_map:
            category_map[i] = category_map[i][0]/category_map[i][1]
        #print(category_map)

        ans['result'] = [(k,v) for k,v in sorted(category_map.items(),key = lambda x:(-x[1],x[0]))[:n]] 
    with open(output,"w") as file:
        file.write(json.dumps(ans,indent=4))

File: 1. Spark Familiarization/submission/task2.scala
import org.json4s.jackson.JsonMethods._
import org.json4s._
import org.json4s.jackson.Serialization.write
import org.json4s.JsonDSL._
import org.apache.spark.sql.SparkSession
import java.io._
import scala.collection.mutable.ListBuffer



object task2 {
  def main(args: Array[String]): Unit = {
    implicit val formats = org.json4s.DefaultFormats
    val spark = args(3)
    val n = args(4).toInt
    var ans = JObject()
    if (spark == "spark") {
      val ss = SparkSession
        .builder()
        .appName("scala")
        .config("spark.master", "local[*]")
        .getOrCreate()
      val sc = ss.sparkContext
      val review = sc.textFile(args(0))
      val business = sc.textFile(args(1))

      val bsn = business.map(x => parse(x)).filter(x => (x \\ "categories") != JNull).map(x => ((x \\ "business_id").values.toString, (x \\ "categories").values.toString.split(','))).flatMap(x => for (n <- x._2) yield { (x._1, n.trim()) })
      val rev = review.map(x => parse(x)).map(x => ((x \\ "business_id").values.toString, (x \\ "stars").values.toString))
      val a = bsn.join(rev).map(x => x._2).aggregateByKey((0.0, 0))((x: (Double, Int), y: String) => (x._1 + y.toFloat, x._2 + 1), (x: (Double, Int), y: (Double, Int)) => (x._1 + y._1, x._2 + y._2)).mapValues(x => x._1 / x._2).sortBy(key => (-key._2, key._1), ascending = true).take(n)
      var k = a.map(t => List(JString(t._1), JDouble(t._2))).toList

      ans = ans ~ ("result", k)
    } else {
      var review = new ListBuffer[List[String]]()
      for (line <- scala.io.Source.fromFile(args(0)).getLines) review += List((parse(line) \\ "business_id").values.toString, (parse(line) \\ "stars").values.toString)

      var business = new ListBuffer[List[String]]()
      for (line <- scala.io.Source.fromFile(args(1)).getLines) {
        if (parse(line) \\ "categories" != JNull) {
          business += List((parse(line) \\ "business_id").values.toString, (parse(line) \\ "categories").values.toString)
        }
      }

      var review_map = scala.collection.mutable.Map[String, Array[Double]]()
      for (r <- review) {
        if (review_map.contains(r(0))) {
          review_map(r(0))(0) = review_map(r(0))(0) + r(1).toDouble
          review_map(r(0))(1) = review_map(r(0))(1) + 1
        } else {
          review_map += (r(0) -> Array(r(1).toDouble, 1))

        }
      }

      var category_map = scala.collection.mutable.Map[String, Array[Double]]()
      for (b <- business) {
        if (review_map.contains(b(0)) & b(0) != JNull) {
          for (j <- b(1).split(',')) {

            if (!category_map.contains(j.trim())) {
              category_map += (j.trim() -> Array(0, 0))
            }
            category_map(j.trim())(0) = category_map(j.trim())(0) + review_map(b(0))(0)
            category_map(j.trim())(1) = category_map(j.trim())(1) + review_map(b(0))(1)
          }
        }
      }

      var ans_map = new ListBuffer[List[Any]]()
      for (i <- category_map.keys) {
        ans_map += List(i,category_map(i)(0) / category_map(i)(1))
      }

      ans = ans ~ ("result", ans_map.sortBy(x=>(-x(1).toString.toDouble,x(0).toString)).slice(0,n).map(t=>List(JString(t(0).toString),JDouble(t(1).toString.toDouble))))

    }

    val file = new File(args(2))
    val output = new BufferedWriter(new FileWriter(file))
    output.write(write(ans))
    output.close()
  }



}
